tests/framework/extensions/kubectl/apply.go

```diff
@@ -0,0 +1,187 @@
+package kubectl
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+
+	apisV1 "github.com/rancher/rancher/pkg/apis/provisioning.cattle.io/v1"
+	"github.com/rancher/rancher/tests/framework/clients/dynamic"
+	"github.com/rancher/rancher/tests/framework/clients/rancher"
+	"github.com/rancher/rancher/tests/framework/extensions/kubeconfig"
+	ext_unstructured "github.com/rancher/rancher/tests/framework/extensions/unstructured"
+	"github.com/rancher/rancher/tests/framework/pkg/wait"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/watch"
+	k8Scheme "k8s.io/client-go/kubernetes/scheme"
+	restclient "k8s.io/client-go/rest"
+
+	management "github.com/rancher/rancher/tests/framework/clients/rancher/generated/management/v3"
+	rbacv1 "k8s.io/api/rbac/v1"
+)
+
+
+// KubectlApply is a helper function that creates a Job and executing kubectl apply by copying the yaml
+// in the pod. It then executes kubectl apply and returns the output from the pod logs. 
+func KubectlApply(client *rancher.Client, cluster *apisV1.Cluster, yamlContent *management.ImportClusterYamlInput, clusterID string) (string, error) {
+
+	var restConfig *restclient.Config
+
+	kubeConfig, err := kubeconfig.GetKubeconfig(client, clusterID)
+	if err != nil {
+		return "", err
+	}
+
+	restConfig, err = (*kubeConfig).ClientConfig()
+	if err != nil {
+		return "", err
+	}
+	restConfig.ContentConfig.NegotiatedSerializer = serializer.NewCodecFactory(k8Scheme.Scheme)
+
+	ts := client.Session.NewSession()
+	defer ts.Cleanup()
+
+	downClient, err := dynamic.NewForConfig(ts, restConfig)
+	if err != nil {
+		return "", err
+	}
+
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "rancher-installer",
+		},
+	}
+	_, err = downClient.Resource(corev1.SchemeGroupVersion.WithResource("serviceaccounts")).Namespace(Namespace).Create(context.TODO(), ext_unstructured.MustToUnstructured(sa), metav1.CreateOptions{})
```

Comment:
Maybe for these kubectl calls, we can consolidated the resource creation into a function, and the apply and explain calls just receive the kubectl call. Also do we need to crate the service accounts and cluster role bindings.

Reply:
yes we need the sa but not cluster role bindings as we have to get the pod logs without the SA, we would get a forbidden error.